src/originalhtml.rs

```diff
@@ -0,0 +1,594 @@
+    let mimetype = ctype.mimetype.to_lowercase();
+    if mimetype.starts_with("multipart") && ctype.params.get("boundary").is_some() {
+        MimeS::Multiple
+    } else if mimetype == "message/rfc822" {
+        MimeS::Message
+    } else {
+        MimeS::Single
+    }
+}
+
+// HtmlMsgParser converts a mime-message to HTML.
+#[derive(Debug)]
+pub struct HtmlMsgParser {
+    pub html: String,
+    pub plain: Option<String>,
+    pub format_flowed: bool,
+    pub delsp: bool,
+}
+
+impl HtmlMsgParser {
+    pub async fn from_bytes(context: &Context, rawmime: &[u8]) -> Result<Self> {
+        let mut parser = HtmlMsgParser {
+            html: "".to_string(),
+            plain: None,
+            format_flowed: false,
+            delsp: false,
+        };
+
+        let parsedmail = mailparse::parse_mail(rawmime)?;
+
+        parser.collect_texts_recursive(context, &parsedmail).await?;
+
+        if parser.html.is_empty() {
+            if let Some(plain) = parser.plain.clone() {
+                parser.html = plain_to_html(&plain, parser.format_flowed, parser.delsp).await;
+            }
+        } else {
+            parser.cid_to_data_recursive(context, &parsedmail).await?;
+        }
+
+        Ok(parser)
+    }
+
+    fn collect_texts_recursive<'a>(
+        &'a mut self,
+        context: &'a Context,
+        mail: &'a mailparse::ParsedMail<'a>,
+    ) -> Pin<Box<dyn Future<Output = Result<bool>> + 'a + Send>> {
+        use futures::future::FutureExt;
+
+        // Boxed future to deal with recursion
+        async move {
+            match get_mimes(&mail.ctype).await {
+                MimeS::Multiple => {
+                    let mut any_part_added = false;
+                    for cur_data in mail.subparts.iter() {
+                        if self.collect_texts_recursive(context, cur_data).await? {
+                            any_part_added = true;
+                        }
+                    }
+                    Ok(any_part_added)
+                }
+                MimeS::Message => {
+                    let raw = mail.get_body_raw()?;
+                    if raw.is_empty() {
+                        return Ok(false);
+                    }
+                    let mail = mailparse::parse_mail(&raw).unwrap();
+                    self.collect_texts_recursive(context, &mail).await
+                }
+                MimeS::Single => {
+                    let mimetype = mail.ctype.mimetype.parse::<Mime>()?;
+                    if mimetype == mime::TEXT_HTML {
+                        if let Ok(decoded_data) = mail.get_body() {
+                            self.html = decoded_data;
+                            return Ok(true);
+                        }
+                    } else if mimetype == mime::TEXT_PLAIN {
+                        if let Ok(decoded_data) = mail.get_body() {
+                            self.plain = Some(decoded_data);
+                            self.format_flowed =
```

Comment:
these parameter-getting lines are similar to the ones from mimeparser.rs, maybe we can move them to format_flowed.rs, however, as not much more code from format_flowed.rs is reusable (format_flowed.rs convets to plain text ignoring quotes where we want to convert to html and regard quotes), i think, it is not worth the noise.